File: frontend/lib/services/danger_zone_service.dart
import 'package:dio/dio.dart';
import '../models/danger_zone.dart';
import 'api_client.dart';

class DangerZoneService {
  Future<List<DangerZone>> fetchAll({
    String bbox = '106.6,10.7,107.0,11.0',
    int limit = 10,
    CancelToken? cancelToken,
  }) async {
    final res = await ApiClient.instance.get(
      '/danger-zones',
      queryParameters: {'bbox': bbox, 'limit': limit},
      cancelToken: cancelToken,
    );
    final list = _extractList(res.data);
    return list
        .whereType<Map>()
        .map((e) => DangerZone.fromJson(Map<String, dynamic>.from(e)))
        .where((zone) => zone.hasPolygon || zone.hasCircle)
        .toList();
  }

  Future<List<DangerZone>> fetchByBounds({
    required double minLat,
    required double minLng,
    required double maxLat,
    required double maxLng,
    int limit = 10,
    CancelToken? cancelToken,
  }) {
    return fetchAll(
      bbox: '$minLng,$minLat,$maxLng,$maxLat',
      limit: limit,
      cancelToken: cancelToken,
    );
  }

  List<dynamic> _extractList(dynamic data) {
    if (data is List) return data;

    if (data is Map<String, dynamic>) {
      if (data['type'] == 'FeatureCollection') {
        return (data['features'] as List?) ?? [];
      }
      return (data['data'] as List?) ?? [];
    }

    return [];
  }

  Future<bool> checkLocation({
    required double lat,
    required double lng,
    CancelToken? cancelToken,
  }) async {
    final res = await ApiClient.instance.get(
      '/danger-zones/check',
      queryParameters: {'lat': lat, 'lng': lng},
      cancelToken: cancelToken,
    );
    if (res.statusCode == 200) {
      final data = res.data['data'];
      if (data != null) return data['isDanger'] ?? data['inside'] ?? false;
    }
    return false;
  }
}

File: frontend-rn/src/services/danger-zone.service.ts
import { apiClient } from './api-client';
import {
  dangerZoneFromJson,
  hasPolygon,
  hasCircle,
  type DangerZone,
} from '@/types/danger-zone';

function extractList(data: any): any[] {
  if (Array.isArray(data)) return data;
  if (data && typeof data === 'object') {
    if (data.type === 'FeatureCollection') return data.features ?? [];
    return data.data ?? [];
  }
  return [];
}

export async function fetchDangerZones(params: {
  bbox?: string;
  limit?: number;
  signal?: AbortSignal;
}): Promise<DangerZone[]> {
  const { bbox = '106.6,10.7,107.0,11.0', limit = 10, signal } = params;
  const res = await apiClient.get('/danger-zones', {
    params: { bbox, limit },
    signal,
  });
  return extractList(res.data)
    .filter((e) => e && typeof e === 'object')
    .map(dangerZoneFromJson)
    .filter((z) => hasPolygon(z) || hasCircle(z));
}

export function fetchDangerZonesByBounds(p: {
  minLat: number;
  minLng: number;
  maxLat: number;
  maxLng: number;
  limit?: number;
  signal?: AbortSignal;
}) {
  return fetchDangerZones({
    bbox: `${p.minLng},${p.minLat},${p.maxLng},${p.maxLat}`,
    limit: p.limit,
    signal: p.signal,
  });
}

export async function checkLocationDanger(p: {
  lat: number;
  lng: number;
  signal?: AbortSignal;
}): Promise<boolean> {
  const res = await apiClient.get('/danger-zones/check', {
    params: { lat: p.lat, lng: p.lng },
    signal: p.signal,
  });
  if (res.status === 200) {
    const d = res.data?.data;
    if (d) return d.isDanger ?? d.inside ?? false;
  }
  return false;
}
